### ISSUE TITLE
webpack issues a HTTP request with wrong chunk URL under certain circumstances

### ISSUE THREAD

**Issue by jsnajdr** [webpack/webpack#11005]
webpack version: 4.42.0

This is a very complex bug: when app has multiple entrypoints, has a separate runtime for each entrypoint, there is an async module that's also loaded statically by one of the entrypoints, and the `splitChunks` optimization decides to lay out shared chunks is a certain way, then the compiled app will:
- issue a HTTP request for the async module although it already has it available statically
- the HTTP request will have incorrect URL: it will use the numeric (natural) ID of the module instead of the chunk name

The result is a runtime error: 404 response from server and the `import()` returns a reject promise with `Loading chunk { id } failed` error.

The expected behavior is that the request is not issued at all.

This is extracted from a large app with thousands of modules ([Calypso](https://github.com/automattic/wp-calypso)) and I don't have a minimalistic reproduction case that would really fail. The trouble with the example I'm presenting below is that I'm unable to trick webpack to split the shared chunks in exactly the way I need.

But I know enough about the bug, and can reason about what's wrong, to warrant a report and discussion:

Entrypoint 1:
```js
import 'react';
import './async';
import './load-async';
```

Entrypoint 2:
```js
import 'react';
import './load-async';
```

Entrypoint 3:
```js
import './load-async';
```

`async` module:
```js
import 'react';
```

Common `load-async` module that dynamically imports the async module:
```js
import(/* webpackChunkName: "async" */ './async');
```

Now, let webpack produce five chunks from this:

- `entry-1`: contains modules `entry-1`, `load-async` and `async`
- `entry-2`: contains modules `entry-2` and `load-async`
- `vendors~entry-1~entry-2`: contains module `react` (shared vendors chunk)
- `entry-3`: contains modules `entry-3` and `load-async`
- `async`: contains modules `async` and `react`

webpack decided to ship `react` twice: once in a shared chunk, and once in the `async` module. Depending on the various optimization thresholds (min/max size, max parallel requests, ...) this can certainly happen.

When building the chunk graph, webpack figures out that the chunk group `entry-1` includes all modules from chunk group `async` (thanks to the static import of `async`), and therefore it doesn't connect these groups: `entry-1` doesn't have `async` as child.

That means that the function `jsonpScriptSrc` generated in the runtime doesn't have a record for the `async` module that would map its numeric ID to the chunk name. If the `async` chunk has id=6, and the file name is `async.js`, then:
```js
__webpack_require__.e(6)
```
will issue HTTP request for file `6.js`, which doesn't exist. That can cause a runtime error: `Loading chunk 6 failed.`

Webpack has an optimization that should prevent this most of the time: the `FlagIncludedChunksPlugin`. That plugin should figure out that `entry-1` is a superset of `async` and push the ID of `async` to `chunk.ids` of  `entry-1`. Then, when generating the runtime for `entry-1`, the `installedChunks` map would contain the ID of `async` from the very beginning, and `__webpack_require__.e` would always detect it as loaded and never issue an actual HTTP request.

Alas, that optimization fails here, because `entry-1` chunk is not a superset of the `async` chunk: the `react` module is not in `entry-1`, but in the shared chunk!

If it compared the chunk groups instead, like `buildChunkGraph` does, then it would discover that chunk group `entry-1` is a superset of chunk group `async`. Comparing chunk groups and chunks yields different results.

There are few necessary conditions to trigger this bug:
- webpack must generate a custom runtime for each entrypoint. If we specified `runtimeChunk: 'single'`, the ID -> name map in `jsonpScriptSrc` would have a record for `async`, because it would be forced by another entrypoint that doesn't import `async` statically.
- the dynamic import must be done in a common module imported in multiple entrypoints. Otherwise, webpack would replace `__webpack_require__.e(6)` with a simple `Promise.resolve()`
- the `entry-1` entrypoint must also include another async module that's not statically available. Other wise, webpack would detect that the `__webpack_require__.e` code is not really used and the implementation would be a stub that returns `Promise.resolve()`

**Comment by sokra** [comment]
I understand the problem. The runtime knows that this chunk does need to be loaded, but the module code

**Comment by sokra** [comment]
This is still a huge problem, but I don't have a good idea to solve that yet.

**Comment by ScriptedAlchemy** [comment]
Is this still a problem under Webpack 5

**Comment by sokra** [comment]
fixed by #11489
